source/server/http/stats_handler.cc

```diff
@@ -190,52 +190,65 @@ uint64_t PrometheusStatsFormatter::statsAsPrometheus(
-    const std::string metric_name = metricName(gauge->tagExtractedName());
-    if (metric_type_tracker.find(metric_name) == metric_type_tracker.end()) {
-      metric_type_tracker.insert(metric_name);
-      response.add(fmt::format("# TYPE {0} gauge\n", metric_name));
-    }
-    response.add(fmt::format("{0}{{{1}}} {2}\n", metric_name, tags, gauge->value()));
-  }
+  // From
+  // https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md#grouping-and-sorting:
+  //
+  // All lines for a given metric must be provided as one single group, with the optional HELP and
+  // TYPE lines first (in no particular order). Beyond that, reproducible sorting in repeated
+  // expositions is preferred but not required, i.e. do not sort if the computational cost is
+  // prohibitive.
+
+  // Processes a metric type (counter, gauge, histogram) by generating all output lines, sorting
+  // them by metric name, and then outputting them in the correct sorted order into response.
+  auto process_type = [&](const auto& metrics, const auto& generate_name_and_output,
+                          absl::string_view type) -> uint64_t {
+    // This is a sorted collection to satisfy the "preferred" ordering from the prometheus
+    // spec: metrics will be sorted by their tags' textual representation, which will be consistent
+    // across calls.
+    std::map<std::string, std::set<std::string>> groups;
+    for (const auto& metric : metrics) {
+      if (!shouldShowMetric(*metric, used_only, regex)) {
+        continue;
+      }
 
-  for (const auto& histogram : histograms) {
-    if (!shouldShowMetric(*histogram, used_only, regex)) {
-      continue;
+      const auto name_output_pair = generate_name_and_output(*metric);
+      groups[name_output_pair.first].emplace(std::move(name_output_pair.second));
     }
 
-    const std::string tags = formattedTags(histogram->tags());
-    const std::string hist_tags = histogram->tags().empty() ? EMPTY_STRING : (tags + ",");
-
-    const std::string metric_name = metricName(histogram->tagExtractedName());
-    if (metric_type_tracker.find(metric_name) == metric_type_tracker.end()) {
-      metric_type_tracker.insert(metric_name);
-      response.add(fmt::format("# TYPE {0} histogram\n", metric_name));
+    for (const auto& group : groups) {
+      response.add(fmt::format("# TYPE {0} {1}\n", group.first, type));
+      for (const auto& output : group.second) {
+        response.add(output);
+      }
+      response.add("\n");
     }
-
-    const Stats::HistogramStatistics& stats = histogram->cumulativeStatistics();
+    return groups.size();
+  };
+
+  // Returns a pair of metric_name and prometheus output line for a counter or a gauge.
+  auto generate_counter_and_gauge_output =
+      [](const auto& metric) -> std::pair<std::string, std::string> {
+    const std::string tags = formattedTags(metric.tags());
+    const std::string metric_name = metricName(metric.tagExtractedName());
```

Comment:
I'm worried about the explosion of string-storage during the sort. Can we populate a vector of metrics, and sort them by Metric::tagExtractedStatName()? StatNames are sortable.

Then we could stream out the formatted value?

I realize that at the moment the admin output infrastructure buffers everything, so I'm not under the delusion we can in this PR have the prometheus handler consume constant space. But I think we could make it consume half the space it consumes with this approach, and that's probably a significant memory bump. And I think it wouldn't be hugely difficult to update the admin output infrastructure to stream rather than fully buffer.

There was a talk at KubeCon by Splunk about stats cardinality blowing through hundreds of gigs of memory in Prometheus [(slides)](https://static.sched.com/hosted_files/kccncna19/2c/great_cardinality_disasters_of_our_time.pdf) and the source of the stats explosion was Istio.

Reply:
also to sort by StatName you can use:
https://github.com/envoyproxy/envoy/blob/62777e87edc9f16af5c7920bb86f198585789a3d/source/common/stats/symbol_table_impl.h#L697

Reply:
I worried about this too. Let me take a pass at this and see how it comes out.